modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/ckeditor.jsp

```diff
@@ -471,252 +472,310 @@ name = HtmlUtil.escapeJS(name);
+
+					if (editorWrapper) {
+						editorWrapper.once('mouseenter', function (event) {
+							ckEditor.once(
+								'focus',
+								onBlur.bind(this, document.activeElement)
+							);
+							ckEditor.focus();
+						});
+					}
+				});
 			}
-		});
+		}
 
-		ckEditor.on('setData', (event) => {
-			instanceDataReady = false;
-		});
+		currentToolbarSet = getToolbarSet(initialToolbarSet);
+
+		var defaultConfig = {
+			filebrowserBrowseUrl: '',
+			filebrowserFlashBrowseUrl: '',
+			filebrowserImageBrowseLinkUrl: '',
+			filebrowserImageBrowseUrl: '',
+			filebrowserUploadUrl: null,
+			toolbar: currentToolbarSet,
+		};
 
-		if (UA.edge && parseInt(UA.edge, 10) >= 14) {
-			var resetActiveElementValidation = function (activeElement) {
-				activeElement = A.one(activeElement);
+		var editorConfig = <%= Validator.isNotNull(editorConfigJSONObject) ? editorConfigJSONObject : "{}" %>;
 
-				var activeElementAncestor = activeElement.ancestor();
+		var config = A.merge(defaultConfig, editorConfig);
 
-				if (
-					activeElementAncestor.hasClass('has-error') ||
-					activeElementAncestor.hasClass('has-success')
-				) {
-					activeElementAncestor.removeClass('has-error');
-					activeElementAncestor.removeClass('has-success');
+		var editorTransformerURLs = config.editorTransformerURLs;
 
-					var formValidatorStack = activeElementAncestor.one(
-						'.form-validator-stack'
-					);
+		if (Liferay.FeatureFlags['LPS-186870'] && editorTransformerURLs) {
+			var loadingIndicator = document.createElement('span');
 
-					if (formValidatorStack) {
-						formValidatorStack.remove();
-					}
-				}
-			};
+			loadingIndicator.classList.add('loading-animation');
+			loadingIndicator.setAttribute('aria-hidden', true);
 
-			var onBlur = function (activeElement) {
-				resetActiveElementValidation(activeElement);
+			editorContainer.appendChild(loadingIndicator);
 
-				setTimeout(() => {
-					if (activeElement) {
-						ckEditor.focusManager.blur(true);
-						activeElement.focus();
-					}
-				}, 0);
-			};
+			Liferay.Util.loadClientExtensions([
+				{
+					clientExtensionDefinitions: editorTransformerURLs.map(
+						(url) => ({
+							importDeclaration: 'default from ' + url,
+						})
+					),
+					onLoad: (bindingContexts) => {
+						var transformedConfig = config;
 
-			ckEditor.on('instanceReady', () => {
-				var editorWrapper = A.one('#cke_<%= name %>');
+						bindingContexts.forEach(
+							({binding: editorTransformer, error}) => {
+								if (error) {
+									console.error(error);
+								}
 
-				if (editorWrapper) {
-					editorWrapper.once('mouseenter', function (event) {
-						ckEditor.once(
-							'focus',
-							onBlur.bind(this, document.activeElement)
+								if (
+									editorTransformer &&
+									editorTransformer.editorConfigTransformer
+								) {
+									transformedConfig = editorTransformer.editorConfigTransformer(
+										transformedConfig
+									);
+								}
+							}
```

Comment:
all this part of the `loadClientExtension()` code looks the [same as the one we did in BaseEditor](https://github.com/liferay-frontend/liferay-portal/pull/3902/commits/7e7f2c99258b2b2991c12852a5c3a75328ac92c1). Would it be possible to reuse it somehow? Maybe a function taking the `config` and returning a promise resolving to another `config` would do to be used from any editor.

Below it, we do different logic to manage the loading indicator and to actually set the `config`.

I'm fine leaving this as it is and abstracting out in a followup task

Reply:
We can revisit this when we apply CETs to Alloy Editor. I am hesitant to introduce a util, or a set of utils, in the scope of `frontend-editor-ckeditor-web`, as our goal it to eventually remove support for the legacy version. Maybe the ideal scope of util is to place it in `frontend-editor-api` or new `frontend-editor-web`? :thinking: